File: src/main/java/io/polygloat/controllers/ImportController.java
package io.polygloat.controllers;

import io.polygloat.constants.Message;
import io.polygloat.dtos.PathDTO;
import io.polygloat.dtos.request.ImportDto;
import io.polygloat.exceptions.NotFoundException;
import io.polygloat.model.Language;
import io.polygloat.model.Permission;
import io.polygloat.model.Repository;
import io.polygloat.model.Source;
import io.polygloat.service.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.validation.SmartValidator;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@CrossOrigin(origins = "*")
@RequestMapping("/api/repository/{repositoryId}/import")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ImportController implements IController {

    private final TranslationService translationService;
    private final SourceService sourceService;
    private final RepositoryService repositoryService;
    private final SecurityService securityService;
    private final LanguageService languageService;
    private final SmartValidator validator;

    @PostMapping(value = "")
    public void doImport(@PathVariable("repositoryId") Long repositoryId, @RequestBody ImportDto dto) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.MANAGE);
        Repository repository = repositoryService.findById(repositoryId).orElseThrow(() -> new NotFoundException(Message.REPOSITORY_NOT_FOUND));

        Language language = languageService.getOrCreate(repository, dto.getLanguageAbbreviation());

        for (Map.Entry<String, String> entry : dto.getData().entrySet()) {
            Source source = sourceService.getOrCreateSource(repository, PathDTO.fromFullPath(entry.getKey()));
            translationService.setTranslation(source, language.getAbbreviation(), entry.getValue());
        }
    }
}

File: src/main/kotlin/io/polygloat/controllers/ImportController.kt
package io.polygloat.controllers

import io.polygloat.constants.Message
import io.polygloat.dtos.ImportDto
import io.polygloat.exceptions.NotFoundException
import io.polygloat.model.Permission
import io.polygloat.service.*
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody
import java.io.OutputStream
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import javax.validation.Valid


@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping("/api/repository/{repositoryId}/import")
class ImportController(
        private val repositoryService: RepositoryService,
        private val securityService: SecurityService,
        private val importService: ImportService
) : IController {

    @PostMapping(value = [""])
    fun doImport(@PathVariable("repositoryId")
                 repositoryId: Long,
                 @RequestBody @Valid
                 dto: ImportDto
    ): ResponseEntity<StreamingResponseBody> {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.MANAGE)
        val repository = repositoryService.findById(repositoryId).orElseThrow {
            NotFoundException(Message.REPOSITORY_NOT_FOUND)
        }

        val stream = StreamingResponseBody { out: OutputStream ->
            importService.import(repository, dto, out)
        }

        return ResponseEntity(stream, HttpStatus.OK)

    }
}